src/qibolab/instruments/qm/devices.py

```diff
@@ -0,0 +1,133 @@
+from collections import defaultdict
+from dataclasses import dataclass, field
+from itertools import chain
+from typing import Dict
+
+from qibolab.instruments.abstract import Instrument
+
+from .ports import (
+    OPXIQ,
+    OctaveInput,
+    OctaveOutput,
+    OPXInput,
+    OPXOutput,
+    QMInput,
+    QMOutput,
+)
+
+
+class PortsDefaultdict(defaultdict):
+    """Dictionary mapping port numbers to
+    :class:`qibolab.instruments.qm.ports.QMPort` objects.
+
+    Automatically instantiates ports that have not yet been created.
+    Used by :class:`qibolab.instruments.qm.devices.QMDevice`
+
+    https://stackoverflow.com/questions/2912231/is-there-a-clever-way-to-pass-the-key-to-defaultdicts-default-factory
+    """
+
+    def __missing__(self, key):
+        ret = self[key] = self.default_factory(key)  # pylint: disable=E1102
+        return ret
+
+
+@dataclass
+class QMDevice(Instrument):
+    """Abstract class for an individual Quantum Machines devices."""
+
+    name: str
+    """Name of the device."""
+
+    outputs: Dict[int, QMOutput] = field(init=False)
+    """Dictionary containing the instrument's output ports."""
+    inputs: Dict[int, QMInput] = field(init=False)
+    """Dictionary containing the instrument's input ports."""
+
+    def ports(self, number, input=False):
+        """Provides instrument's ports to the user.
+
+        Args:
+            number (int): Port number.
+                Can be 1 to 10 for :class:`qibolab.instruments.qm.devices.OPXplus`
+                and 1 to 5 for :class:`qibolab.instruments.qm.devices.Octave`.
+            input (bool): ``True`` for obtaining an input port, otherwise an
+                output port is returned. Default is ``False``.
+        """
+        if input:
+            return self.inputs[number]
+        else:
+            return self.outputs[number]
```

Comment:
Just style. 

In principle, I could tell you to omit the `else` because of the `return`. But there is a fancier way
```suggestion
        ports_ = self.inputs if input else self.outputs
        return ports_[number]
```

Reply:
932725c2572c55dcd25f2168fd3582a10f04cbf4